python/ray/test/cluster_utils.py

```diff
@@ -0,0 +1,181 @@
+                services.get_node_ip_address(),
+                self.redis_address,
+                cleanup=True,
+                use_raylet=True,
+                **kwargs)
+            # TODO(rliaw): Find a more stable way than modifying global state.
+            process_dict_copy = services.all_processes.copy()
+            for key in services.all_processes:
+                services.all_processes[key] = []
+            node = Node(process_dict_copy)
+            self.worker_nodes[node] = address_info
+        logging.info("Starting Node with raylet socket {}".format(
+            address_info["raylet_socket_names"]))
+
+        return node
+
+    def remove_node(self, node):
+        """Kills all processes associated with worker node.
+
+        Args:
+            node (Node): Worker node of which all associated processes
+                will be removed.
+        """
+        if self.head_node == node:
+            self.head_node.kill_allprocesses()
+            self.head_node = None
+            # TODO(rliaw): Do we need to kill all worker processes?
+        else:
+            node.kill_allprocesses()
+            self.worker_nodes.pop(node)
+
+        assert not node.any_processes_alive(), (
+            "There are zombie processes left over after killing...")
+
+    def wait_for_nodes(self):
+        """Waits for all nodes to be registered with global state."""
+        try:
+            while True:
+                registered = len([
+                    client for client in ray.global_state.client_table()
+                    if client["IsInsertion"]
+                ])
+                expected = len(self.list_all_nodes())
+                if registered == expected:
+                    logger.info("All nodes registered as expected!")
+                    break
+                else:
+                    logger.info(
+                        "Currently registering {} but expecting {}".format(
+                            registered, expected))
+                time.sleep(0.3)
+        except Exception as e:
+            logger.exception("Perhaps try Cluster(connect=True)?")
+
+    def list_all_nodes(self):
+        """Lists all nodes.
+
+        TODO(rliaw): What is the desired behavior if a head node
+        dies before worker nodes die?
+
+        Returns:
+            List of all nodes, including the head node."""
+        nodes = list(self.worker_nodes)
+        if self.head_node:
+            nodes = [self.head_node] + nodes
+        return nodes
+
+    def shutdown(self):
+        # We create a list here as a copy because `remove_node`
+        # modifies `self.worker_nodes`.
+        all_nodes = list(self.worker_nodes)
+        for node in all_nodes:
+            self.remove_node(node)
+        self.remove_node(self.head_node)
+
+
+class Node():
+    """Abstraction for a Ray node."""
+
+    def __init__(self, process_dict):
+        # TODO(rliaw): Is there a unique identifier for a node?
```

Comment:
Each node has a `client_id` in the client table. What kind of identifier are you looking for?

Reply:
Yeah; something like that. This is non-blocking; what would be a way to get the client_id without first calling `ray.init`? Ideally, it would be part of the output of `services.start_ray_node`.